### PR TITLE
Added LazyBsonDocument and made SinkDocument key/values lazily decoded.

### PR DESCRIPTION
**Pull request by rozza**
KAFKA-45

https://evergreen.mongodb.com/version/5e84aa5a7742ae7105ed9ccd

### CODE REVIEW

File: src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    @Override
+    public BsonValue remove(final Object key) {
+        return getUnwrapped().remove(key);
+    }
+
+    @Override
+    public void putAll(final Map<? extends String, ? extends BsonValue> m) {
+        super.putAll(m);
```
Comment (jyemin) [inline src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java]: Probably unused, but should this also be `getUnwrapped().putAll(m)`?

File: src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
```
Comment (jyemin) [inline src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java]: Same: `getUnwrapped().clear()`?

**Review by jyemin — COMMENTED** [review]
It looks like it does what is claims to do.  I'm not super clear on the user scenario under which this would come in to play though, so it might be helpful to provide that context.

File: src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java (hunk 2 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java]: copy / pasta will update

File: src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java (hunk 1 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java]: copy / pasta will update

**Comment by rozza** [comment]
Updated.  For context: Sink Records have both a Key value and a Value value. 

Most the time the connector only cares about the value part and ignores the key part. However, at present if the key is not convertible into a document it will fail.  This requires the user to do an extra Kafka configuration step to make the key convertible (via a Single Message Transform). It also throws a cryptic error to the user about what to do.

Often keys are used as a partitioning values for Kafka - so are message metadata that isn't stored by the sink connector at all. 

In short this change makes life easier for users.

**Review by jyemin — APPROVED** [review]
LGTM!
